### ISSUE TITLE
Soft links in zip distribution

### ISSUE THREAD

**Issue by dridzal** [trilinos/Trilinos#1012]
I received an email from a collaborator at LANL, who downloaded the zip file from the github website.  Apparently, his build fails due to soft links in CMakeLists files.  I'm not sure I fully understand the issue, but here is the explanation:

_In the zip distribution, the CMakeLists.txt file for the poisson-inversion test was a soft-link, but it pointed to some kind of long regex list, apparently describing the lines they wanted in the file.  I wasn't familiar with that type of soft-link, but I figure it's something the developers somehow use temporarily when they're testing something new.  But it broke this configuration script, so I just replaced it with a regular CMakeLists.txt file for the two tests in that directory and it then it worked fine.  The tests even ran and passed.  You might want to pass this on to whoever controls the distribution, so they can update the zip file. The last time I tried to clone something from git-hub it didn't work because of our firewall.  Maybe there's a workaround, but I don't know it, because I usually only use our local one._

Have you seen this?

**Comment by bmpersc** [comment]
@dridzal to be able to address this I'm going to need more information

It looks like "zip distribution" is one of the download options github provides, is that correct? If so what "version" was downloaded? In general we don't really support those downloads. However, it should just be a copy of the source from a specific sha1 which should effectively be the same as having a clone at the same sha1.

Which CMakeLists.txt is being discussed here?

Is the regex comment related to the symlink issue or separate? These are two very different concepts that don't typically interact. From reading this it sounds like the symlink is a regex of the files that it points to, but that isn't how symlinks work so that can't be what is meant.

I'm not sure what we can do about the firewall issue cloning from github. However, since this person was able to download something from github it seems like the https connections aren't firewalled. They could trying cloning using https instead of ssh. Maybe that will work better?

**Comment by dridzal** [comment]
I think this is the zip file that you get as an alternative to git clone.  You can try it out by clicking on the 'Clone or download' link.

The file is in /Trilinos/packages/rol/example/poisson-inversion.

I've given him some input on how to set the proxy.  I'm waiting to hear back.  Still, I'm not sure that it is wise to ignore the zip file as a means of distribution.  We can't control how our potential users obtain the code.  In any case, the copies should be identical, and, apparently, they are not.

**Comment by bmpersc** [comment]
@dridzal I'm not able to replicate this. The repo and the download of the master branch have the CMakeLists.txt in packages/rol/example/poisson-inversion as a real file. I can't even find a symlink in ROL period. I also can't find a CMakeLists.txt that is a symlink in all of Trilinos.

At this point I don't know what to suggest. Nothing plausible comes to mind how it would be possible for your customer have such a wildly different copy of Trilinos that what we have short of there being serious issues with their disk or computer in general. However, that seems highly unlikely that such a failure would produce rational, but wrong data. Without more information, such as what is the regex that was mentioned I don't know how I can help with this.

**Comment by dridzal** [comment]
I can confirm this with my own download.  Maybe the zip/unzip utility and the platform matter ... still there should be no symlinks.  I'll close the issue

**Comment by tawiesn** [comment]
I had problems with the downloaded zip, too. I tried the standard "unzip" in a CentOS 7 container and the resulting unpacked files were partially screwed up. Not all packages were affected though and i could not find the reason for it. Then i tried pk7zip to unpack the zip file and everything was fine.

**Comment by dridzal** [comment]
This is unfortunate.  It may be best to disable the 'download zip' link on github.  @bmpersc , can we do that?

**Comment by bmpersc** [comment]
Hmm, my experiment was to use the standard unzip on the mac and I'm getting exactly what I expect. I know mac and linux tools are not the same, but I find it very disturbing that the linux version is improperly decompressing things. @tawiesn was your "partially screwed up" a case of seemingly random data changes or was it reasonable looking stuff that was just wrong? I could understand the former happening if the (de)compression method used was wrong, but it should affect all files (though for this to happen and pass the parity checks is highly unlikely).

@dridzal I don't believe that we can disable that download. I know I checked previously for a way to disable the release downloads and couldn't find one, but that was a year ago. Maybe somethings have changed and we can control that now. I will look into it, but rarely do we have control over the github UI. It might also be worthwhile to report this as a bug to github. I'm sure they don't intend for the zip download option to behave this way.

**Comment by tawiesn** [comment]
@bmpersc When using the standard "unzip" it looked good, but single files in some packages were not unpacked correctly (missing/screwed lines, wrong EOF,...). I can't remember the details, but it was unusable for compiling. By googling around i found a hint, that a different unzip utility could work. And pk7zip did the job for me. But finally i decided to use git clone over https. If we could deactivate the download zip button we certainly would do a lot of users a favor... At least we should officially announce that git clone is the preferred way to get the source code.

**Comment by bmpersc** [comment]
Just to satisfy my curiosity I decided to unzip the zip I'd downloaded on a couple linux machines. I ended up with the same result. The CMakeLists.txt in question was a real file in all instances. In case I suffered from a general garbling issue I built MueLu on one of the linux machines and was able to successfully compile it. So I think it is safe to say that the zip I have is good. I don't really know why I would have a good zip while others have broken ones, but I think that goes back to there may just be a bug on github's side.

From my googling and searching through options on github I don't believe there is a way to disable these downloads. I recognize that this can be confusing for our users, but in absence of a way to control this there is little we can do to prevent people from using these github provided tools and getting themselves in trouble.
